core/state_processor.go

```diff
@@ -98,23 +101,50 @@ func (p *StateProcessor) Process(block *types.Block, statedb *state.StateDB, cfg
 		allLogs = append(allLogs, receipt.Logs...)
 	}
 
+	// Overlay tree migration logic
+	migrdb := statedb.Database()
+	// TODO: avoid opening the preimages file here and make it part of, potentially, statedb.Database().
+	filePreimages, err := os.Open("preimages.bin")
+	if err != nil {
+		return nil, nil, 0, fmt.Errorf("opening preimage file: %s", err)
+	}
+	defer filePreimages.Close()
+	preimageSeek := migrdb.GetCurrentPreimageOffset()
+	if _, err := filePreimages.Seek(migrdb.GetCurrentPreimageOffset(), io.SeekStart); err != nil {
+		return nil, nil, 0, fmt.Errorf("seeking preimage file: %s", err)
+	}
+	fpreimages := bufio.NewReader(filePreimages)
```

Comment:
First thing we do after opening the `preimage.bin` file, is to seek the last saved position in the file, which is always _the next_ preimage that should be read.